pkg/logs/internal/decoder/auto_multiline_detection/tokenizer.go

```diff
@@ -0,0 +1,387 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package automultilinedetection contains auto multiline detection and aggregation logic.
+package automultilinedetection
+
+import (
+	"bytes"
+	"unicode"
+)
+
+// Token is the type that represents a single token.
+type Token byte
+
+// maxRun is the maximum run of a char or digit before it is capped.
+// Note: This must not exceed d10 or c10 below.
+const maxRun = 10
+
+const (
+	space Token = iota
+
+	// Special Characters
+	colon        // :
+	semicolon    // ;
+	dash         // -
+	underscore   // _
+	fslash       // /
+	bslash       // \
+	period       // .
+	comma        // ,
+	singlequote  // '
+	doublequote  // "
+	backtick     // `
+	tilda        // ~
+	star         // *
+	plus         // +
+	equal        // =
+	parenopen    // (
+	parenclose   // )
+	braceopen    // {
+	braceclose   // }
+	bracketopen  // [
+	bracketclose // ]
+	ampersand    // &
+	exclamation  // !
+	at           // @
+	pound        // #
+	dollar       // $
+	percent      // %
+	uparrow      // ^
+
+	// Digit runs
+	d1
+	d2
+	d3
+	d4
+	d5
+	d6
+	d7
+	d8
+	d9
+	d10
+
+	// Char runs
+	c1
+	c2
+	c3
+	c4
+	c5
+	c6
+	c7
+	c8
+	c9
+	c10
+
+	// Special tokens
+	month
+	day
+	apm // am or pm
+	zone
+	t
```

Comment:
Why `t` is a special token?

Reply:
To improve accuracy when matching and detecting timestamps in formats like this `2024-03-28T14:33:53.743350Z`

Reply:
I added a comment here for clarification